domain/consensus/processes/pruningmanager/pruningmanager.go

```diff
@@ -1,51 +1,226 @@
+		pruningDepth:          pruningDepth(uint64(k), finalityInterval, constants.MergeSetSizeLimit),
+		finalityInterval:      finalityInterval,
 	}
 }
 
 // FindNextPruningPoint finds the next pruning point from the
 // given blockHash
-func (pm *pruningManager) FindNextPruningPoint(blockHash *externalapi.DomainHash) error {
-	return nil
+func (pm *pruningManager) FindNextPruningPoint() error {
+	virtual, err := pm.ghostdagDataStore.Get(pm.databaseContext, model.VirtualBlockHash)
+	if err != nil {
+		return err
+	}
+
+	currentP, err := pm.PruningPoint()
+	if err != nil {
+		return err
+	}
+	currentPGhost, err := pm.ghostdagDataStore.Get(pm.databaseContext, currentP)
+	if err != nil {
+		return err
+	}
+	currentPBlueScore := currentPGhost.BlueScore
+	// Because the pruning point changes only once per finality, then there's no need to even check for that if a finality interval hasn't passed.
+	if virtual.BlueScore <= currentPBlueScore+pm.finalityInterval {
+		return nil
+	}
+
+	// This means the pruning point is still genesis.
+	if virtual.BlueScore <= pm.pruningDepth+pm.finalityInterval {
+		return nil
+	}
+
+	// get Virtual(pruningDepth)
+	candidatePHash, err := pm.dagTraversalManager.HighestChainBlockBelowBlueScore(model.VirtualBlockHash, pm.pruningDepth)
+	if err != nil {
+		return err
+	}
+	candidatePGhost, err := pm.ghostdagDataStore.Get(pm.databaseContext, candidatePHash)
+	if err != nil {
+		return err
+	}
+
+	// Actually check if the pruning point changed
+	if (currentPBlueScore / pm.finalityInterval) < (candidatePGhost.BlueScore / pm.finalityInterval) {
+		utxoIter, err := pm.consensusStateManager.RestorePastUTXOSetIterator(candidatePHash)
+		if err != nil {
+			return err
+		}
+		serializedUtxo, err := serializeUTXOSetIterator(utxoIter)
+		if err != nil {
+			return err
+		}
+		pm.pruningStore.Stage(candidatePHash, serializedUtxo)
+		currentP = candidatePHash
+	}
+	return pm.deletePastBlocks(currentP)
 }
 
 // PruningPoint returns the hash of the current pruning point
 func (pm *pruningManager) PruningPoint() (*externalapi.DomainHash, error) {
-	return nil, nil
+	pruningPoint, err := pm.pruningStore.PruningPoint(pm.databaseContext)
+	if err != nil {
+		if !pm.pruningStore.HasPruningPoint(pm.databaseContext) {
+			genesis, err := pm.dagTraversalManager.HighestChainBlockBelowBlueScore(model.VirtualBlockHash, pm.pruningDepth+pm.finalityInterval)
+			if err != nil {
+				return nil, err
+			}
+			utxoGenesisIter, err := pm.consensusStateManager.RestorePastUTXOSetIterator(genesis)
+			if err != nil {
+				return nil, err
+			}
+			serializedUtxo, err := serializeUTXOSetIterator(utxoGenesisIter)
+			if err != nil {
+				return nil, err
+			}
+			pm.pruningStore.Stage(genesis, serializedUtxo)
+			pruningPoint = genesis
+		} else {
+			return nil, err
+		}
+
+	}
+	return pruningPoint, nil
 }
 
 // SerializedUTXOSet returns the serialized UTXO set of the
 // current pruning point
 func (pm *pruningManager) SerializedUTXOSet() ([]byte, error) {
```

Comment:
This function is no longer relevant, as anyone who needs the serialized UTXOSet can simply directly get it from the store. Please remove it.